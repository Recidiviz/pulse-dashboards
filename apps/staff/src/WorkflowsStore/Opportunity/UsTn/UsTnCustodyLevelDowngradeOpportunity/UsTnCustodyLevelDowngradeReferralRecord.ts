// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { cloneDeep } from "lodash";
import { z } from "zod";

import { opportunitySchemaBase } from "~datatypes";

import { formInformationSchema as formInformation } from "../UsTnSharedCriteria";

export const usTnCustodyLevelDowngradeSchema = opportunitySchemaBase
  .extend({
    eligibleCriteria: z.object({
      custodyLevelHigherThanRecommended: z.object({
        custodyLevel: z.string(),
        recommendedCustodyLevel: z.string(),
      }),
      custodyLevelIsNotMax: z.null(),
      usTnLatestCafAssessmentNotOverride: z.object({
        overrideReason: z.string().nullable(),
      }),
      usTnIneligibleForAnnualReclassification: z.object({
        ineligibleCriteria: z.array(z.string()),
      }),
    }),
    formInformation,
  })
  .transform((r) => {
    const out = cloneDeep(r);
    if (!out.caseNotes["ASSAULTIVE DISCIPLINARIES"])
      out.caseNotes["ASSAULTIVE DISCIPLINARIES"] = [];
    return out;
  });

export type UsTnCustodyLevelDowngradeReferralRecordRaw = z.input<
  typeof usTnCustodyLevelDowngradeSchema
>;

export type UsTnCustodyLevelDowngradeReferralRecord = z.infer<
  typeof usTnCustodyLevelDowngradeSchema
>;
