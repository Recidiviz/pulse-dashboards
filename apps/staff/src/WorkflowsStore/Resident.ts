// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { addYears } from "date-fns";
import { uniqBy } from "lodash";

import { PortionServedDates, WorkflowsResidentRecord } from "../FirestoreStore";
import { ResidentMetadata } from "../FirestoreStore/types";
import { RootStore } from "../RootStore";
import tenants from "../tenants";
import { JusticeInvolvedPersonBase } from "./JusticeInvolvedPersonBase";
import {
  IncarcerationOpportunityType,
  Opportunity,
  OpportunityFactory,
  UsMeSCCPOpportunity,
} from "./Opportunity";
import { UsIdExpandedCRCOpportunity } from "./Opportunity/UsId";
import { UsIdCRCResidentWorkerOpportunity } from "./Opportunity/UsId/UsIdCRCResidentWorkerOpportunity/UsIdCRCResidentWorkerOpportunity";
import { UsIdCRCWorkReleaseOpportunity } from "./Opportunity/UsId/UsIdCRCWorkReleaseOpportunity";
import {
  UsMeFurloughReleaseOpportunity,
  UsMeWorkReleaseOpportunity,
} from "./Opportunity/UsMe";
import { UsMeAnnualReclassificationOpportunity } from "./Opportunity/UsMe/UsMeAnnualReclassificationOpportunity";
import { UsMeMediumTrusteeOpportunity } from "./Opportunity/UsMe/UsMeMediumTrusteeOpportunity";
import { usMiAddInPersonSecurityClassificationCommitteeReviewOpportunity } from "./Opportunity/UsMi/UsMiAddInPersonSecurityClassificationCommitteeReviewOpportunity";
import { usMiReclassificationRequestOpportunity } from "./Opportunity/UsMi/UsMiReclassificationRequestOpportunity";
import { usMiSecurityClassificationCommitteeReviewOpportunity } from "./Opportunity/UsMi/UsMiSecurityClassificationCommitteeReviewOpportunity";
import { usMiWardenInPersonSecurityClassificationCommitteeReviewOpportunity } from "./Opportunity/UsMi/UsMiWardenInPersonSecurityClassificationCommitteeReviewOpportunity";
import { UsMoOverdueRestrictiveHousingInitialHearingOpportunity } from "./Opportunity/UsMo/UsMoOverdueRestrictiveHousingInitialHearingOpportunity";
import { UsMoOverdueRestrictiveHousingReleaseOpportunity } from "./Opportunity/UsMo/UsMoOverdueRestrictiveHousingReleaseOpportunity";
import { UsMoOverdueRestrictiveHousingReviewHearingOpportunity } from "./Opportunity/UsMo/UsMoOverdueRestrictiveHousingReviewHearingOpportunity";
import { UsTnCustodyLevelDowngradeOpportunity } from "./Opportunity/UsTn";
import { UsTnAnnualReclassificationReviewOpportunity } from "./Opportunity/UsTn/UsTnAnnualReclassificationReviewOpportunity/UsTnAnnualReclassificationReviewOpportunity";
import { fractionalDateBetweenTwoDates, optionalFieldToDate } from "./utils";

const LIFE_SENTENCE_THRESHOLD = addYears(new Date(), 200);

const residentialOpportunityConstructors: Partial<
  Record<
    IncarcerationOpportunityType,
    new (c: Resident) => Opportunity<Resident>
  >
> = {
  usIdCRCResidentWorker: UsIdCRCResidentWorkerOpportunity,
  usIdCRCWorkRelease: UsIdCRCWorkReleaseOpportunity,
  usIdExpandedCRC: UsIdExpandedCRCOpportunity,
  usMeFurloughRelease: UsMeFurloughReleaseOpportunity,
  usMeMediumTrustee: UsMeMediumTrusteeOpportunity,
  usMeSCCP: UsMeSCCPOpportunity,
  usMeWorkRelease: UsMeWorkReleaseOpportunity,
  usTnCustodyLevelDowngrade: UsTnCustodyLevelDowngradeOpportunity,
  usTnAnnualReclassification: UsTnAnnualReclassificationReviewOpportunity,
  usMoOverdueRestrictiveHousingRelease:
    UsMoOverdueRestrictiveHousingReleaseOpportunity,
  usMoOverdueRestrictiveHousingInitialHearing:
    UsMoOverdueRestrictiveHousingInitialHearingOpportunity,
  usMoOverdueRestrictiveHousingReviewHearing:
    UsMoOverdueRestrictiveHousingReviewHearingOpportunity,
  usMeReclassificationReview: UsMeAnnualReclassificationOpportunity,
  usMiReclassificationRequest: usMiReclassificationRequestOpportunity,
  usMiSecurityClassificationCommitteeReview:
    usMiSecurityClassificationCommitteeReviewOpportunity,
  usMiWardenInPersonSecurityClassificationCommitteeReview:
    usMiWardenInPersonSecurityClassificationCommitteeReviewOpportunity,
  usMiAddInPersonSecurityClassificationCommitteeReview:
    usMiAddInPersonSecurityClassificationCommitteeReviewOpportunity,
};

const createResidentOpportunity: OpportunityFactory<
  IncarcerationOpportunityType,
  Resident
> = (type, person) => {
  const constructor = residentialOpportunityConstructors[type];
  if (constructor) return new constructor(person);
};

export class Resident extends JusticeInvolvedPersonBase<WorkflowsResidentRecord> {
  constructor(record: WorkflowsResidentRecord, rootStore: RootStore) {
    super(record, rootStore, createResidentOpportunity);
  }

  get facilityId(): string | undefined {
    return this.record.facilityId ?? undefined;
  }

  get gender(): string | undefined {
    return this.record.gender;
  }

  get unitId(): string | undefined {
    return this.record.unitId ?? undefined;
  }

  get custodyLevel(): string | undefined {
    return this.record.custodyLevel ?? undefined;
  }

  get admissionDate(): Date | undefined {
    return optionalFieldToDate(this.record.admissionDate);
  }

  get releaseDate(): Date | undefined {
    return optionalFieldToDate(this.record.releaseDate);
  }

  get onLifeSentence(): boolean | undefined {
    const { releaseDate } = this.record;

    if (!releaseDate) return;

    return new Date(releaseDate) > LIFE_SENTENCE_THRESHOLD;
  }

  get searchIdValue(): any {
    const { currentTenantId } = this.rootStore;
    const searchField =
      currentTenantId &&
      tenants[currentTenantId]?.workflowsSystemConfigs?.INCARCERATION
        ?.searchField;

    return searchField ? this.record[searchField] : this.assignedStaffId;
  }

  get sccpEligibilityDate(): Date | undefined {
    if (this.record.metadata.stateCode !== "US_ME") {
      return;
    }

    return optionalFieldToDate(this.record.metadata.sccpEligibilityDate);
  }

  get usTnFacilityAdmissionDate(): Date | undefined {
    return optionalFieldToDate(this.record.usTnFacilityAdmissionDate);
  }

  get usNdParoleReviewDate(): Date | undefined {
    if (this.record.metadata.stateCode !== "US_ND") {
      return;
    }

    return optionalFieldToDate(this.record.metadata.paroleReviewDate);
  }

  get metadata(): ResidentMetadata {
    return this.record.metadata;
  }

  get portionServedDates(): PortionServedDates {
    const startDate = optionalFieldToDate(this.record.admissionDate);
    const endDate = optionalFieldToDate(this.record.releaseDate);
    const halfTimeDate = fractionalDateBetweenTwoDates(startDate, endDate, 0.5);
    const twoThirdsTimeDate = fractionalDateBetweenTwoDates(
      startDate,
      endDate,
      2 / 3,
    );

    const opportunityDates: PortionServedDates = [];

    if (this.metadata.stateCode === "US_ME") {
      opportunityDates.push({
        heading: "Half Time",
        date: halfTimeDate,
      });

      if (this.metadata.portionServedNeeded === "2/3")
        opportunityDates.push({
          heading: "Two Thirds Time",
          date: twoThirdsTimeDate,
        });
    }

    return uniqBy(opportunityDates, "heading");
  }
}
