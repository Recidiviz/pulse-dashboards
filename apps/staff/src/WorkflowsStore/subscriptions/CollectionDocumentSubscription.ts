// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { doc, DocumentData, DocumentReference } from "firebase/firestore";

import FirestoreStore from "../../FirestoreStore";
import { FIRESTORE_COLLECTIONS_MAP } from "../../FirestoreStore/constants";
import type { FirestoreCollectionKey } from "../../FirestoreStore/types";
import { FirestoreDocumentSubscription } from "./FirestoreDocumentSubscription";
import { TransformFunction, UpdateFunction, ValidateFunction } from "./types";

/**
 * Subscribes to the specified document in the specified collection.
 * RecordType is essentially just an assertion for downstream convenience;
 * the data source itself is not validated or even strongly typed. Proceed with caution!
 */
export class CollectionDocumentSubscription<
  RecordType extends DocumentData
> extends FirestoreDocumentSubscription<RecordType> {
  readonly dataSource: DocumentReference;

  constructor(
    firestoreStore: FirestoreStore,
    firestoreCollectionKey: FirestoreCollectionKey,
    recordId: string,
    transformFn?: TransformFunction<RecordType>,
    validateFn?: ValidateFunction<RecordType>,
    updateFn?: UpdateFunction<DocumentData>
  ) {
    super(transformFn, validateFn, updateFn);

    this.dataSource = doc(
      firestoreStore.db,
      FIRESTORE_COLLECTIONS_MAP[firestoreCollectionKey],
      recordId
    );
  }
}
