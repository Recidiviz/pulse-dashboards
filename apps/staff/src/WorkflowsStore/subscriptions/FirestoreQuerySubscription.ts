// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import * as Sentry from "@sentry/react";
import {
  DocumentData,
  onSnapshot,
  Query,
  QuerySnapshot,
  Unsubscribe,
} from "firebase/firestore";
import {
  action,
  computed,
  makeObservable,
  override,
  reaction,
  runInAction,
} from "mobx";
import { IDisposer } from "mobx-utils";

import { FeatureGateError } from "../../errors";
import { FirestoreSubscription } from "./FirestoreSubscription";
import {
  QuerySubscription,
  TransformFunction,
  ValidateFunction,
} from "./types";
import { defaultTransformFunction, defaultValidateFunction } from "./utils";

export abstract class FirestoreQuerySubscription<
    DataFormat extends DocumentData,
  >
  extends FirestoreSubscription<DataFormat[]>
  implements QuerySubscription<DataFormat>
{
  abstract get dataSource(): Query | undefined;

  protected disposeDynamicDataSource?: IDisposer;

  constructor(
    protected transformRecord: TransformFunction<DataFormat> = defaultTransformFunction,
    protected validateRecord: ValidateFunction<DataFormat> = defaultValidateFunction,
  ) {
    super([]);

    makeObservable<this, "updateData">(this, {
      dataSource: computed,
      updateData: action,
      subscribe: override,
      unsubscribe: override,
    });
  }

  protected updateData(snapshot: QuerySnapshot | undefined): void {
    const docs: DataFormat[] = [];

    // receiving no snapshot at all is a special case that indicates there is no valid data source;
    // when this happens we want to reset the hydration state in addition to clearing out any data
    // that may already exist from previous updates
    if (!snapshot) {
      this.data = docs;
      this.hydrationState = { status: "needs hydration" };
      return;
    }

    const errors: unknown[] = [];

    snapshot.forEach((doc) => {
      try {
        const record = this.transformRecord(
          doc.data({ serverTimestamps: "estimate" }),
        );

        if (record !== undefined) this.validateRecord(record);

        if (record !== undefined) {
          docs.push(record);
        }
      } catch (e) {
        // Move on to allow for partial success
        errors.push(e);
        // don't log routine feature flag checks, but do log everything else
        if (!(e instanceof FeatureGateError)) {
          Sentry.captureException(e);
        }
      }
    });

    this.data = docs;
    // regardless of errors we always consider this hydrated
    this.hydrationState = { status: "hydrated" };
  }

  protected startSnapshotListener(): Unsubscribe | undefined {
    if (!this.dataSource) {
      this.updateData(undefined);
    } else {
      return onSnapshot(
        this.dataSource,
        (result) => this.updateData(result),
        (error) => {
          runInAction(() => {
            // an error that occurs after data is fetched is considered a partial failure,
            // but this one means that the fetch operation itself failed and therefore so does hydration.
            this.hydrationState = { status: "failed", error };
            this.data = [];
          });
          Sentry.captureException(error);
        },
      );
    }
  }

  subscribe(): void {
    super.subscribe();

    // watches dataSource for changes and re-subscribes accordingly
    this.disposeDynamicDataSource = reaction(
      () => this.dataSource,
      () => {
        this.unsubscribe();
        this.subscribe();
      },
    );
  }

  unsubscribe(): void {
    super.unsubscribe();
    this.disposeDynamicDataSource?.();
  }
}
