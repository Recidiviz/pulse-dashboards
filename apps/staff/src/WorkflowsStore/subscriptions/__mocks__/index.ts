// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { MockDocumentSubscription } from "./MockDocumentSubscription";
import { MockQuerySubscription } from "./MockQuerySubscription";

export const CollectionDocumentSubscription = MockDocumentSubscription;

export const OpportunityUpdateSubscription = MockDocumentSubscription;

export const CaseloadSubscription = MockQuerySubscription;

export const StaffSubscription = MockQuerySubscription;

export const UserSubscription = MockQuerySubscription;

export const LocationSubscription = MockQuerySubscription;
