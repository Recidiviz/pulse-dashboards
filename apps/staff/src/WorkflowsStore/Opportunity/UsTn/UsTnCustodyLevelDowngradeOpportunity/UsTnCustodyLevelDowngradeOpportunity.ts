// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { DocumentData } from "firebase/firestore";

import { OpportunityUpdateWithForm } from "../../../../FirestoreStore";
import { Resident } from "../../../Resident";
import { UsTnReclassificationReviewForm } from "../../Forms/UsTnReclassificationReviewForm";
import { OpportunityBase } from "../../OpportunityBase";
import { UsTnSharedReclassificationDraftData } from "../UsTnSharedCriteria";
import {
  UsTnCustodyLevelDowngradeReferralRecord,
  usTnCustodyLevelDowngradeSchema,
} from "./UsTnCustodyLevelDowngradeReferralRecord";

export class UsTnCustodyLevelDowngradeOpportunity extends OpportunityBase<
  Resident,
  UsTnCustodyLevelDowngradeReferralRecord,
  OpportunityUpdateWithForm<UsTnSharedReclassificationDraftData>
> {
  form: UsTnReclassificationReviewForm;

  // TODO(#6707) move to configuration
  readonly caseNotesTitle = "Relevant Information For Classification";

  constructor(resident: Resident, record: DocumentData) {
    super(
      resident,
      "usTnCustodyLevelDowngrade",
      resident.rootStore,
      usTnCustodyLevelDowngradeSchema.parse(record),
    );

    this.form = new UsTnReclassificationReviewForm(this, resident.rootStore);
  }
}
