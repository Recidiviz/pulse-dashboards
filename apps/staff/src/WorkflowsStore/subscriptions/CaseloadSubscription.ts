// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import {
  DocumentData,
  FieldPath,
  Query,
  query,
  where,
} from "firebase/firestore";
import { z } from "zod";

import { FirestoreCollectionKey } from "~firestore-api";

import { WorkflowsStore } from "../WorkflowsStore";
import { FirestoreQuerySubscription } from "./FirestoreQuerySubscription";

export class CaseloadSubscription<
  RecordType extends DocumentData,
> extends FirestoreQuerySubscription<RecordType> {
  workflowsStore: WorkflowsStore;

  firestoreCollectionKey: FirestoreCollectionKey;

  personType: RecordType["personType"];

  constructor(
    workflowsStore: WorkflowsStore,
    firestoreCollectionKey: FirestoreCollectionKey,
    personType: RecordType["personType"],
    parser?: z.ZodType<RecordType, any, any>,
  ) {
    parser ? super(parser.parse) : super();
    this.workflowsStore = workflowsStore;
    this.firestoreCollectionKey = firestoreCollectionKey;
    this.personType = personType;
  }

  get dataSource(): Query | undefined {
    const {
      selectedSearchIds,
      rootStore: { currentTenantId, firestoreStore },
    } = this.workflowsStore;

    if (!currentTenantId || !selectedSearchIds.length) {
      return undefined;
    }

    const { firestoreCollectionKey, personType } = this;

    const config = this.workflowsStore.systemConfigFor(
      personType === "RESIDENT" ? "INCARCERATION" : "SUPERVISION",
    );

    const constraints = [
      where("stateCode", "==", currentTenantId),
      where(
        new FieldPath(...config.searchField),
        config.searchOp ?? "in",
        selectedSearchIds,
      ),
    ];

    if (config.onlySurfaceEligible) {
      constraints.push(where("allEligibleOpportunities", "!=", []));
    }

    return query(
      firestoreStore.collection(firestoreCollectionKey),
      ...constraints,
    ).withConverter({
      fromFirestore(snapshot, options) {
        const doc = snapshot.data(options);
        return { ...doc, recordId: snapshot.id, personType };
      },
      // these collections are read-only, so this should never be used, but it is required by Firestore
      toFirestore(record: any) {
        throw new Error(`Writing to ${firestoreCollectionKey} is not allowed`);
      },
    });
  }
}
