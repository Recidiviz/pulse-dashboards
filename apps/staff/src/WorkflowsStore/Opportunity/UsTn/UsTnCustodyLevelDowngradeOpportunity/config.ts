// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { OpportunityConfig } from "../../OpportunityConfigs";
import { UsTnCustodyLevelDowngradeOpportunity } from "./UsTnCustodyLevelDowngradeOpportunity";

export const usTnCustodyLevelDowngradeConfig: OpportunityConfig<UsTnCustodyLevelDowngradeOpportunity> =
  {
    systemType: "INCARCERATION",
    stateCode: "US_TN",
    urlSection: "custodyLevelDowngrade",
    label: "Custody Level Downgrade",
    dynamicEligibilityText:
      "resident[|s] may be eligible for a custody level downgrade",
    callToAction: "Review and update custody levels.",
    subheading:
      "This alert helps staff identify residents who may be at a higher custody level than recommended and directs staff to complete & submit new classification paperwork based on the resident's latest CAF score.",
    firestoreCollection: "US_TN-custodyLevelDowngradeReferrals",
    snooze: {
      defaultSnoozeDays: 30,
      maxSnoozeDays: 90,
    },
    sidebarComponents: [
      "Incarceration",
      "CaseNotes",
      "UsTnCommonlyUsedOverrideCodes",
    ],
    denialReasons: {
      Other: "Please specify a reason",
    },
    methodologyUrl: "",
    eligibleCriteriaCopy: {
      custodyLevelHigherThanRecommended: {
        text: "Custody level is higher than latest CAF score suggests",
      },
      custodyLevelIsNotMax: { text: "Custody level is not maximum" },
      usTnIneligibleForAnnualReclassification: {
        text: "Not eligible for annual reclassification",
      },
      usTnLatestCafAssessmentNotOverride: {
        text: "Last assessment did not include an override",
      },
    },
    homepagePosition: 2,
  };
