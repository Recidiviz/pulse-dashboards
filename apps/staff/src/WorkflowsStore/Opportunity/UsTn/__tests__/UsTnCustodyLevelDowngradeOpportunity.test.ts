// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { DocumentData } from "firebase/firestore";
import { configure } from "mobx";
import tk from "timekeeper";

import { RootStore } from "../../../../RootStore";
import { Resident } from "../../../Resident";
import { DocumentSubscription } from "../../../subscriptions";
import {
  UsTnCustodyLevelDowngradeEligibleResidentRecord,
  UsTnCustodyLevelDowngradeReferralRecordFixture,
} from "../__fixtures__";
import { UsTnCustodyLevelDowngradeOpportunity } from "../UsTnCustodyLevelDowngradeOpportunity";

vi.mock("../../../subscriptions");

let opp: UsTnCustodyLevelDowngradeOpportunity;
let resident: Resident;
let root: RootStore;
let updatesSub: DocumentSubscription<any>;

function createTestUnit(
  residentRecord: typeof UsTnCustodyLevelDowngradeEligibleResidentRecord,
  opportunityRecord: DocumentData,
) {
  root = new RootStore();
  root.workflowsRootStore.opportunityConfigurationStore.mockHydrated();
  vi.spyOn(root.workflowsStore, "opportunityTypes", "get").mockReturnValue([
    "usTnCustodyLevelDowngrade",
  ]);
  resident = new Resident(residentRecord, root);

  opp = new UsTnCustodyLevelDowngradeOpportunity(resident, opportunityRecord);
}

beforeEach(() => {
  // this lets us spy on observables, e.g. computed getters
  configure({ safeDescriptors: false });
  tk.freeze(new Date(2024, 7, 1));
});

afterEach(() => {
  vi.resetAllMocks();
  tk.reset();
  configure({ safeDescriptors: true });
});

describe("fully eligible resident", () => {
  beforeEach(() => {
    createTestUnit(
      UsTnCustodyLevelDowngradeEligibleResidentRecord,
      UsTnCustodyLevelDowngradeReferralRecordFixture,
    );

    updatesSub = opp.updatesSubscription;
    updatesSub.hydrationState = { status: "hydrated" };
  });

  test("requirements almost met", () => {
    expect(opp.requirementsAlmostMet).toEqual([]);
  });

  test("requirements met", () => {
    expect(opp.requirementsMet).toMatchSnapshot();
  });
});
