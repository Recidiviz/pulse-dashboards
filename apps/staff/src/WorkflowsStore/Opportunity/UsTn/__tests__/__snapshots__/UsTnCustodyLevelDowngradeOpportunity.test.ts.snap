// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fully eligible resident > requirements met 1`] = `
[
  {
    "text": "Custody level is higher than latest CAF score suggests",
    "tooltip": undefined,
  },
  {
    "text": "Custody level is not maximum",
    "tooltip": undefined,
  },
  {
    "text": "Not eligible for annual reclassification",
    "tooltip": undefined,
  },
  {
    "text": "Last assessment did not include an override",
    "tooltip": undefined,
  },
]
`;
